text/0000-ember-data-improved-ajax.md

```diff
@@ -0,0 +1,207 @@
+[#3099](https://github.com/emberjs/data/pull/3099) was to address the lack of
+public API for customizing a request. Unfortunately, it hasn't gone through a
+proper RFC process :pensive:. The feature is not enabled and available in a
+`release`, so this RFC should flesh out the specific API for the feature before
+it can be enabled eventually.
+
+The result of this RFC should make sure that the new hooks allow developers to
+customize a request with only public API.
+
+# Detailed design
+
+An adapter operation is split up into 2 steps
+
+- 1) collecting the properties for the request, i.e. `_requestFor`
+- 2) executing the request, i.e. `makeRequest`
+
+So for a `findRecord` it would look like this:
+
+``` js
+findRecord(store, type, id, snapshot) {
+  let requestOptions = this._requestFor({
+    store, type, id, snapshot,
+    requestType: "findRecord"
+  });
+
+  return RSVP.resolve(requestOptions).then((options) => this.makeRequest(options) );
+}
+```
+
+### `_requestFor`
+
+Each hook for step 1) is invoked with all the arguments which are passed to the
+corresponding adapter method. Additionally `requestType` is set so the hooks
+know for which specific adapter operation they are invoked. To allow for
+asynchronous configurations (e.g. pre-flight request to refresh an
+authentication token), `_requestFor` returns a promise which resolves with an
+object describing the request to make. The `_requestFor` method is private and
+calls the new public hooks.
+
+``` js
+_requestFor(params) {
+  let method  = this.methodForRequest(params);
+  let url     = this.urlForRequest(params);
+  let headers = this.headersForRequest(params);
+  let body    = this.bodyForRequest(params);
+
+  return RSVP.hash({ method, url, headers, body });
+}
+```
+##### `methodForRequest`
+
+Invoked to get the HTTP method for the request. Should return one of `"GET"`,
+`"PUT"`, `"POST"` or `"DELETE"`.
+
+##### `urlForRequest`
+
+Get the URL for the request. This hook returns the URL for the request which
+should be made, e.g. `/users/1` or `api.example.com:4321/users?name=abc`
+
+This hook uses 2 new public sub-hooks to construct the URL:
+
+- `pathForRequest` - re-uses the existing `urlForXXX` hooks
+- `queryForRequest` - get an object for all query params for the request
+
+The existing [`sortQueryParams`](http://emberjs.com/api/data/classes/DS.RESTAdapter.html#method_sortQueryParams)
+hook is reused within the new `queryForRequest`. An warning/deprecation should
+be logged when the path already contains query params, as the new
+`queryForRequest` hook should be used for that.
+
+Pseudo implementation:
+
+``` js
+urlForRequest(options) {
+  let path = this.pathForRequest(options);
+  let query = this.queryForRequest(options);
+
+  if (path.indexOf('?') !== -1) {
+    warning("use queryForRequest to specify query params");
+  } else if (query) {
+    let sortedQueryParams = this.sortQueryParams(query);
+    let queryParams = Ember.$.param(sortedQueryParams);
```

Comment:
Ah, arrays and nested queries are definitely a valid use case 👍 

Deprecating `sortQueryParams` in favor of `queryStringForRequest` sounds reasonable to me!